### ISSUE TITLE
What about including InfoZip?

### ISSUE THREAD

**Issue by acailly** [jart/cosmopolitan#189]
This is a real question, I don't know enough to say if it is a good idea or a completely dumb one :smile: 

This project is marvel: crossplatform web server, include Lua, is a self modifying archive and I saw that you plan to include Sqlite :heart_eyes: 

The part I struggled the most in the beginning is the archive part (see #184), and my conclusion was: use InfoZip, it works every time :tada: 
So now I can get a redbean exe, put some assets in it and here is my crossplatform webapp
But if I want to develop on several env, say Linux and Windows, I have to get InfoZip exe on each platform

So I thought maybe it is possible to build InfoZip from source?

And maybe it is possible to go further, and this is where I have no idea if or how it it feasible, what about including InfoZip so that we can do something like `redbean.com zip redbean.com hello.html` ?

**Comment by jart** [comment]
It's totally doable to make an APE binary out of InfoZip's `zip` and `unzip` commands. It could be as simple as saying `gcc *.c ...`  in the appropriate InfoZip folder, provided of course you pass the recommended flags in the README, as well as `-isystem cosmopolitan/libc/isystem` for compatibility. That's using the amalgamated release.

Another option is this project has a third_party folder. If you go through the commit history, and look at the process we've followed in the past where we've added other projects using the two-commit style, then I would accept a pull request that does that for InfoZip. That will give people an even easier path to building their own APE InfoZip. They would just git clone cosmopolitan and then `make o//third_party/infozip/zip.com`. 

Please note, if we do that, I'm not interested at this time in distributing the zip.com binary myself on the website, since there currently isn't an intent to be in the distributor business. I would prefer that we convince the InfoZip authors of APE's merits and help them improve their build process so that they can be the ones publishing that zip.com binary.

**Comment by jart** [comment]
As for `redbean.com zip redbean.com hello.html` one Lua function you might find interesting is [`StoreAsset`](https://justine.lol/redbean/#StoreAsset) which effectively does the same thing as the `zip` command, except with POSIX advisory locks, rather than atomic replacement.

**Comment by acailly** [comment]
Thank you for your answer @jart 
I really like the pragmatic approach you take, not willing to make it an all purpose project impossible to maintain :smiley: 

If I find the time to do so, I will try to make an APE binary of InfoZip, but I'm not fluent with development in C

Regarding `StoreAssets`, I've seen it but since it is `currently only supported on Linux, XNU, and FreeBSD` that doesn't work if I have one of my dev env on Windows

**Comment by jart** [comment]
Re-opening because I realized our Makefile config currently depends on InfoZip, so we have legitimate need of vendoring it in the codebase, in the interest of hermeticity.

We'll get `StoreAsset` working on Windows OpenBSD, and NetBSD soon.

Concerning pragmatism, projects like GNU, Linux, Apache, etc. are already doing a good job at being a big tent. Rather than do the same thing, this project has managed to fill a niche where we've created a tent for the other 80% of use cases that's simple elegant and beautiful.

**Comment by acailly** [comment]
Nice ! thanks :pray: 
(did you forget to actually reopen the issue?)

**Comment by ahgamut** [comment]
The `zip` utility from Info-ZIP compiles with the appropriate `zip.mk` recipe.

Source code from [here](http://infozip.sourceforge.net/Zip.html). Sample screenshot:

![2021-06-10_06-46-32_1366x768](https://user-images.githubusercontent.com/41098605/121449630-d1bdb100-c9b7-11eb-9c2e-a8f65755b9ba.png)

@jart I can put together a PR for `third_party/infozip/zip.com`. Should I add the recipes for `zipnote`, `zipsplit`, and `zipcloak` in `zip.mk`? 

On a related note, I was trying to compile [GNU Make](http://ftpmirror.gnu.org/make/) with cosmopolitan: running `make.com` at the root of the cosmopolitan repo builds everything (on Debian Linux), but `make.com -j4` doesn't work, possibly because I messed with some flags. I'll open a separate issue for this later.

**Comment by jart** [comment]
@ahgamut Sounds good to me. If `make -j4` doesn't work then maybe it's using `clone()`? I'd be interested in learning more.

**Comment by ahgamut** [comment]
A couple of details:

1. In `third_party/infozip/zip.mk`, can I add `-isystem libc/isystem` to the compilation flags instead of replacing the headers in each of the source files?
2. in #162, some `.h` files were changed to `.inc` because they failed a `.h.ok` check. What parameters should I pass to `make o//third_party/infozip/zip.com` so that the build fails if the `.h.ok` check fails?

**Comment by jart** [comment]
1. As you wish. In that case I might iterate on your change because I'll likely want it for my IDE tooling e.g. `make tags` which is confused by angle brackets. It's also a big picture build performance thing. When I profiled TensorFlow builds years back, I noticed that 50% of GCC's time was spent doing `stat()` system calls because TF used lots of `-isystem` paths. Headers are quadratic so having more than one include root linearly raises that quadratic cost in terms of iops. 

2. Header well-formedness is less of a concern here, since we would never link `zip` and `unzip` are programs that we wouldn't use as a library like we do with sqlite. So I would recommend not defining `THIRD_PARTY_INFOZIP_HDRS`.

For future reference, here's the general pattern to follow for doing header well-formedness tests.

```make
THIRD_PARTY_SQLITE3_A_HDRS = $(filter %.h,$(THIRD_PARTY_SQLITE3_A_FILES))
# ...
THIRD_PARTY_SQLITE3_A_CHECKS = \
        $(THIRD_PARTY_SQLITE3_A_HDRS:%=o/$(MODE)/%.ok) \
        ...
# ...
.PHONY: o/$(MODE)/third_party/sqlite3
o/$(MODE)/third_party/sqlite3: \
        $(THIRD_PARTY_SQLITE3_CHECKS) \
        ...
```

Which reduces to:

```make
THIRD_PARTY_SQLITE3_A_HDRS = third_party/sqlite3/sqlite3.h third_party/sqlite3/sqlite3ext.h
# ...
THIRD_PARTY_SQLITE3_A_CHECKS = \
        o//third_party/sqlite3/sqlite3.h.ok o//third_party/sqlite3/sqlite3ext.h.ok
# ...
.PHONY: o/$(MODE)/third_party/sqlite3
o/$(MODE)/third_party/sqlite3: \
        o//third_party/sqlite3/sqlite3.h.ok o//third_party/sqlite3/sqlite3ext.h.ok
```

Which is triggered by the `make → make all → make o → make o//third_party` hierarchy and matched by the build rule:

```
o/%.h.ok: %.h
        cc -xc -o $@ $<
```

Which compiles the `.h` file as though it were a `.c` file, in order to confirm that it's declared its dependencies.

**Comment by pkulchenko** [comment]
@jart, should this ticket be closed now, since #191 has been merged?
